src/Avalonia.Base/Media/GeometryBuilder.cs

```diff
@@ -0,0 +1,610 @@
+﻿// Portions of this source file are adapted from the Windows Presentation Foundation (WPF) project.
+// (https://github.com/dotnet/wpf)
+//
+// Licensed to The Avalonia Project under the MIT License, courtesy of The .NET Foundation.
+//
+// Portions of this source file are adapted from the WinUI project.
+// (https://github.com/microsoft/microsoft-ui-xaml/tree/winui3/main)
+//
+// Licensed to The Avalonia Project under the MIT License.
+
+// Ignore Spelling: keypoints
+
+using System;
+using System.Runtime.CompilerServices;
+using Avalonia.Utilities;
+
+namespace Avalonia.Media
+{
+    /// <summary>
+    /// Contains internal helpers used to build and draw various geometries.
+    /// </summary>
+    internal class GeometryBuilder
+    {
+        private const double PiOver2 = 1.57079633; // 90 deg to rad
+        private const double Epsilon = 0.00000153; // Same as LayoutHelper.LayoutEpsilon
+
+        /// <summary>
+        /// Draws a new rounded rectangle within the given geometry context.
+        /// Warning: The caller must manage and dispose the <see cref="StreamGeometryContext"/> externally.
+        /// </summary>
+        /// <remarks>
+        /// WinUI: https://github.com/microsoft/microsoft-ui-xaml/blob/93742a178db8f625ba9299f62c21f656e0b195ad/dxaml/xcp/core/core/elements/geometry.cpp#L1072-L1079
+        /// </remarks>
+        /// <param name="context">The geometry context to draw into.</param>
+        /// <param name="keypoints">The rounded rectangle keypoints defining the rectangle to draw.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void DrawRoundedCornersRectangle(
```

Comment:
I wonder why this method has AggressiveInlining. It probably won't be much efficient, since these methods are rather big.

I am going to try to run this PR and review in detail on the next week.

Reply:
Yea, I'm not sure why that's there either at this point. I'll remove it.